### PR TITLE
LanceDB integration update

### PR DESCRIPTION
**Pull request by raghavdixit99**
Added : 

- [x] relevance search (w/wo scores)
- [x] maximal marginal search
- [x] image ingestion
- [x] filtering support
- [x] hybrid search w reranking 

make test, lint_diff and format checked.

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/lancedb.py [hunk 1]
```diff
@@ -227,60 +512,118 @@ def similarity_search(
...
-        connection: Any = None,
+        connection: Optional[Any] = None,
         vector_key: Optional[str] = "vector",
         id_key: Optional[str] = "id",
         text_key: Optional[str] = "text",
         table_name: Optional[str] = "vectorstore",
+        api_key: Optional[str] = None,
+        region: Optional[str] = None,
+        mode: Optional[str] = "overwrite",
+        distance: Optional[str] = "l2",
```
Comment (isahers1) [inline libs/community/langchain_community/vectorstores/lancedb.py:624]: Do these need to be optional?

File: libs/community/langchain_community/vectorstores/lancedb.py [hunk 2]
```diff
@@ -55,17 +66,34 @@ def __init__(
         api_key: Optional[str] = None,
         region: Optional[str] = None,
         mode: Optional[str] = "overwrite",
+        table: Optional[Any] = None,
```
Comment (efriis) [inline libs/community/langchain_community/vectorstores/lancedb.py:69]:
```suggestion
        *,
        table: Optional[Any] = None,
```
make them kwarg-only

File: docs/docs/integrations/vectorstores/lancedb.ipynb
Comment (efriis) [inline docs/docs/integrations/vectorstores/lancedb.ipynb:1]: can we change the pip install to just be an upgrade? Why does it need to be 0.6.13? `pip install -U lancedb`

**Review by efriis — COMMENTED** [review]
what is `tantivy` for?

File: libs/community/langchain_community/vectorstores/lancedb.py [hunk 3]
```diff
@@ -96,6 +124,52 @@ def __init__(
...
+        elif "_relevance_score" in columns:
+            score_col = "_relevance_score"
+        else:
+            score_col = None
+
+        if score_col is None or not score:
+            return [
+                Document(
+                    page_content=results[self._text_key][idx].as_py(),
+                    metadata=results["metadata"][idx].as_py(),
```
Comment (isahers1) [inline libs/community/langchain_community/vectorstores/lancedb.py:158]: Should this be results[idx][self._text_key/"metadata"] instead?

Comment (isahers1) [reply libs/community/langchain_community/vectorstores/lancedb.py:158]: Same comment for 166-167, I might just be missing something

File: libs/community/langchain_community/vectorstores/lancedb.py (hunk 2 above)
Comment (raghavdixit99) [reply libs/community/langchain_community/vectorstores/lancedb.py:69]: do you mean to make the cloud connection components kwarg only or all arguments as kwargs ? any specific reason why ?
thanks

File: libs/community/langchain_community/vectorstores/lancedb.py (hunk 3 above)
Comment (raghavdixit99) [reply libs/community/langchain_community/vectorstores/lancedb.py:158]: Hi, its the arrow convention of slicing,  the results are a `pyarrow.Table` thats why `col_name` needs to precede idx

File: docs/docs/integrations/vectorstores/lancedb.ipynb
Comment (raghavdixit99) [reply docs/docs/integrations/vectorstores/lancedb.ipynb:1]: Yes sure, updating it thanks

File: libs/community/langchain_community/vectorstores/lancedb.py (hunk 1 above)
Comment (raghavdixit99) [reply libs/community/langchain_community/vectorstores/lancedb.py:624]: The api_key, region would be optional as they are only if user wants to use cloud, I change the mode/distance args to be fixed

**Comment by raghavdixit99** [comment]
> what is `tantivy` for?

It is for FTS/Hybrid search

**Pull request merged**